File: 57.InsertInterval/Solution.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        i = 0
        head, end = newInterval
        output = []
        
        while i < len(intervals):
            cur_head, cur_end = intervals[i]
            # we only need to insert one interval
            # and the first task is to find where to start
            if head <= cur_end:
                if end < cur_head:
                    # easy case, new interval shall be at position 0
                    break
                else:
                    # start merging, we will keep doing this
                    # until the updated end < cur_head
                    head = min(head, cur_head)
                    end = max(end, cur_end)
            # merge not start yet, append
            else:
                output.append(intervals[i])
            i += 1
        output.append([head,end])
        output.extend(intervals[i:])
        return output
            
            
            

File: 57.InsertInterval/Solution.cs
public class Solution {
    // Notice this is not the best practical solution
    // Shall come back later for better coding
    // e.g. use linkedlist
    public int[][] Insert(int[][] intervals, int[] newInterval) {
        int head = newInterval[0];
        int end = newInterval[1];
        List<int[]> output = new List<int[]>();
        
        int i = 0;
        while (i < intervals.Count())
        {
            int cur_head = intervals[i][0], cur_end = intervals[i][1];
            if (head <= cur_end)
            {
                if (end < cur_head)
                {
                    break;
                }
                else
                {
                    head = Math.Min(head, cur_head);
                    end = Math.Max(end, cur_end);
                }
            }
            else
            {
                output.Add(intervals[i]);
            }
            i += 1;
        }
        output.Add(new int[2] {head, end});
        
        while (i < intervals.Count())
        {
            output.Add(intervals[i]);
            i++;
        }
        return output.ToArray();
    }
}